File: advent_of_code/2022/01/day01.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func part1(input string) int {
	res := 0
	sum := 0
	for _, line := range strings.Split(input, "\n") {
		line = strings.TrimSpace(line)
		if len(line) == 0 {
			res = max(res, sum)
			sum = 0
			continue
		}

		n, err := strconv.Atoi(line)
		if err != nil {
			panic(err)
		}

		sum += n
	}
	res = max(sum, res)
	return res
}

func part2(input string) int {
	var sums []int
	sum := 0
	for _, line := range strings.Split(input, "\n") {
		line = strings.TrimSpace(line)
		if len(line) == 0 {
			sums = append(sums, sum)
			sum = 0
			continue
		}

		n, err := strconv.Atoi(line)
		if err != nil {
			panic(err)
		}

		sum += n
	}

	sums = append(sums, sum)
	sort.Slice(sums, func(i, j int) bool {
		return sums[i] > sums[j]
	})

	sum = 0
	for i := 0; i < 3; i++ {
		sum += sums[i]
	}
	return sum
}

func main() {
	filename := "input.txt"
	test_input := `1000
	2000
	3000
	
	4000
	
	5000
	6000
	
	7000
	8000
	9000
	
	10000`
	var input string
	const DEBUG = false
	if DEBUG {
		input = test_input
	} else {
		s, err := os.ReadFile(filename)
		if err != nil {
			panic(err)
		}
		input = string(s)
	}

	fmt.Printf("2022 Day 1, part 1: %v\n", part1(input))
	fmt.Printf("2022 Day 1, part 2: %v\n", part2(input))
}

File: ruby/aoc/aoc2016/day01.rb
require 'set'
f=File.read('./input/2016/day01.txt')
d={n:[0,1],s:[0,-1],e:[1,0],w:[-1,0]}
l={n: :w,w: :s,s: :e,e: :n}
r={n: :e,e: :s,s: :w,w: :n}
t={R:r,L:l}

# Part 1
z=:n
a,b=f.split(', ').map{|s|c,n=s[0].to_sym,s[1..].to_i;z=t[c][z];x,y=*d[z];[x*n,y*n]}.reduce([0,0]){|a,b|[a[0]+b[0],a[1]+b[1]]}
p a.abs+b.abs

# Part 2
z=:n;v=Set[];res=nil
f.split(', ')
.map{|s|c,n=s[0].to_sym,s[1..].to_i;z=t[c][z];[d[z]]*n}
.reduce([]){|a,b|a+b}
.reduce([0,0]){|a,b|v.include?(a)?(res=a;break):(c=[a[0]+b[0],a[1]+b[1]];v.add(a);c)}
p res[0].abs+res[1].abs